ds/beis_indicators/trademarks/make_trademarks.py

```diff
@@ -0,0 +1,253 @@
+    '''
+    This function takes the df with lats and lons and labels with NUTS based 
+    on their publication year
+
+    Args:
+        df (df) is a dataframe with longitudes and latitudes
+        tm_year (str) is the date variable we use to geocode
+        tm_threshold (int) is the minimum year
+        method (str) is whether we are creating the indicator using a time 
+        consistent approach 
+            (each year in its NUTS category) or using the latest nuts
+    '''
+    df_2 = df.copy()
+
+    #Drop missing years
+    df_2.dropna(axis=0,subset=[tm_year],inplace=True)
+
+    #Create year
+    df_2['year'] = [int(x.split('-')[0]) for x in df_2[tm_year]]
+
+    #Focus on years after the threshold
+    df_2 = df_2.loc[df_2['year']>=tm_threshold]
+
+    #Create unique postcodee df for reverse geocoding. Here we want to 
+    #remove duplicates to speed things up
+    
+    pc_lookup = df_2.drop_duplicates('pcds_1st')[['pcds_1st','long','lat']]
+
+    #We create a reverse geocoding lookup
+    pc_nuts_map = make_reverse_geocode_lookup(pc_lookup,nuts_years=[2010,2013,2016])
+
+    if method == 'time_consistent':
+        df_2['nuts_code'] = [pc_nuts_map[row['pcds_1st']][get_nuts_category(
+                                row['year'])] 
+            if row['pcds_1st'] in pc_nuts_map.keys() else np.nan for rid,row in df_2.iterrows()]
+    else:
+        df_2['nuts_code'] = [pc_nuts_map[row['pcds_1st']]['nuts2_2016'] if 
+            row['pcds_1st'] in pc_nuts_map.keys() else np.nan for rid,row in df_2.iterrows()]
+
+    return(df_2)
+
+
+def subset_trademark_category(tdf,selected_classes):
+    '''
+    Takes the trademark df and subsets it to focus on a few classes
+
+    Args:
+        tdf (df) is the trademarks df
+        selected_classes (list) are the classes we are interested in
+
+    '''
+    sel_df = tdf.loc[tdf[selected_classes].sum(axis=1)>0]
+    return(sel_df)
+
+def make_trademarks(URL,detailed_classes):
+    """
+    Function to make the trademark dataset. It collects the data and geocodes it.
+
+    Args:
+        URL (str) is the link where we collect the trademark data
+        detailed_classes (dict) is a dict where the keys are names of 
+        #detailed classes and the values lists of class codes
+    """
+    # Download and process the trademark data
+    logger.info("Downloading the data")
+    tm_df = get_trademarks(URL)
+
+    logger.info("Geocoding the data")
+    tm_df_geo = geocode_trademarks(tm_df)
+
+    logger.info("Reverse geocoding the data")
+    tm_df_nuts = reverse_geocode_trademarks(tm_df_geo)
+
+    print(tm_df_nuts.head())
+
+    for x in tm_df_nuts.columns:
+        if "class" in x:
+            tm_df_nuts[x] = pd.to_numeric(tm_df_nuts[x])
+
+    #Create trademarks
+    logger.info("Creating indicators")
+    all_trademarks = tm_df_nuts.groupby(['nuts_code','year']).size()
+    all_trademarks.name = 'total_trademarks'
+
+    make_indicator(all_trademarks,
```

Comment:
it would be great it we could tackle #185 first (it's now used in 8 notebooks and 3 scripts (probably overlapping with the correspondent notebook)) so it should be easy/quick enough? This would avoid us to come back to these scripts again.

Reply:
I'll do this when I come back to the rest of the indicators later in the week.